content/200-orm/200-prisma-client/150-using-raw-sql/100-typedsql.mdx

```diff
@@ -0,0 +1,204 @@
+
+1. In your SQL file, use placeholders for the parameters you want to pass. The syntax for placeholders depends on your database engine:
+
+   <TabbedContent>
+   <TabItem value="PostgreSQL">
+   For PostgreSQL, use the positional placeholders `$1`, `$2`, etc.:
+
+   ```sql title="prisma/sql/getUsersByAge.sql"
+   SELECT id, name, age
+   FROM users
+   WHERE age > $1 AND age < $2
+   ```
+   </TabItem>
+   <TabItem value="MySQL">
+   For MySQL, use the positional placeholders `?`:
+
+   ```sql title="prisma/sql/getUsersByAge.sql"
+   SELECT id, name, age
+   FROM users
+   WHERE age > ? AND age < ?
+   ```
+   </TabItem>
+   <TabItem value="SQLite">
+   In SQLite, there are a number of different placeholders you can use. For this example, we'll use named placeholders `:minAge` and `:maxAge`:
+
+   ```sql title="prisma/sql/getUsersByAge.sql"
+   SELECT id, name, age
+   FROM users
+   WHERE age > :minAge AND age < :maxAge
+   ```
+   </TabItem>
+   </TabbedContent>
+
+   :::note
+   
+   See below for information on how to type placeholders in your SQL files.
+
+   :::
+
+1. When using the generated function in your TypeScript code, pass the arguments as additional parameters to `$queryRawTyped`:
+
+   ```typescript
+   import { PrismaClient } from '@prisma/client'
+   import { getUsersByAge } from '@prisma/client/sql'
+
+   const prisma = new PrismaClient()
+
+   const minAge = 18
+   const maxAge = 30
+   const users = await prisma.$queryRawTyped(getUsersByAge(minAge, maxAge))
+   console.log(users)
+   ```
+
+By using parameterized queries, you ensure type safety and protect against SQL injection vulnerabilities. The TypedSQL generator will create the appropriate TypeScript types for the parameters based on your SQL query, providing full type checking for both the query results and the input parameters.
+
+### Passing array arguments to TypedSQL
+
+TypedSQL supports passing arrays as arguments for PostgreSQL. Use PostgreSQL's `ANY` operator with an array parameter.
+
+```sql title="prisma/sql/getUsersByIds.sql"
+SELECT id, name, email
+FROM users
+WHERE id = ANY($1)
+```
+
+```typescript
+import { PrismaClient } from '@prisma/client'
+import { getUsersByIds } from '@prisma/client/sql'
+
+const prisma = new PrismaClient()
+
+const userIds = [1, 2, 3]
+const users = await prisma.$queryRawTyped(getUsersByIds(userIds))
+console.log(users)
+```
+
+TypedSQL will generate the appropriate TypeScript types for the array parameter, ensuring type safety for both the input and the query results.
+
+:::note
+
+When passing array arguments, be mindful of the maximum number of placeholders your database supports in a single query. For very large arrays, you may need to split the query into multiple smaller queries.
+
+:::
+
+### Defining argument types in your SQL files
+
+Argument typing in TypedSQL is accomplished via specific comments in your SQL files. These comments are of the form:
+
+```sql
+-- @param {Type} $N:alias optional description
```

Comment:
is optional description 2 items or 1 item? I wonder if people will think its 2 items

Reply:
Oh nvm you kinda answer that below.